pkg/provisioner/crdprovisioner.go

```diff
@@ -465,6 +458,99 @@ func (c *CrdProvisioner) PublishVolume(
 		_, err = azVAClient.Create(ctx, azVolumeAttachment, metav1.CreateOptions{})
 		if err != nil {
 			klog.Errorf("Error creating azvolume attachment for volume id (%s) to node id (%s) error : %v", volumeID, nodeID, err)
+		}
+		return publishContext, err
+	}
+	var updateFunc func(obj interface{}) error
+	// if attachment is scheduled for deletion, new attachment should only be created after the deletion
+	if attachmentObj.DeletionTimestamp != nil {
```

Comment:
Shouldn't we make the deletion and pending operation checks before potentially unpublishing other replica attachments?

Reply:
The idea was that if a replica AzVolumeAttachment A gets requested for deletion and another replica B gets created by `replica controller` to replace it, there can be a moment where there are `len(replicaAzVolumeAttachments) =- maxMountReplicaCount + 2`, including a primary AzVolumeAttachment C that got demoted from unpublish request, in which case it made more sense to make a deletion request for C before checking whether A's deletionTimestamp is set as otherwise, we would end up requesting deletion of C only after A's been successfully deleted.

However, I can also see other scenarios where this order wouldn't make sense. The above deletion logic needs a revision to prevent these scenarios.